[FLINK-10977][table] Add UnBounded FlatAggregate operator to streaming Table API

hequn8128 opened the pull request:
## What is the purpose of the change

This pull request adds UnBounded FlatAggregate operator to streaming Table API.
The design doc can be found [here](https://docs.google.com/document/d/1tnpxg31EQz2-MEzSotwFzqatsB4rNLz0I-l_vPa5H4Q/edit#heading=h.q23rny2iglsr).
Some discussion about flatAggregate can be found [here](http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Table-API-Enhancement-Outline-td25070.html).


## Brief change log

  - Add flatAggregate api to Table. In order to close the "flatAggregate" with a select statement, add a FlatAggTable and a GroupedFlatAggTable class.
  - Add TableAggregateFunction for User-Defined Table Aggregates. User can register a TableAggregateFunction through table environment.
  - Add logical relnodes for TableAggregate: `LogicalTableAggregate`, `FlinkLogicalTableAggregate`, `DataStreamTableAggregate`.
  - Add Rules for TableAggregate relnodes.
  - Refactor codegen logic for Aggregations. Add a `TableAggregationCodeGenerator` code generator to generate the TableAggregate runtime function.
  - Add documents and tests.
  - Add a enforceKeyFields() method in `UpsertStreamTableSink` in order to upsert output results of flatAggregate, i.e, `UpsertStreamTableSink` requires keys when upsert data. The enforceKeyFields() return empty array by default. Users can override the method if the data contain logical key fields. However, I find that java subclasses still have to implement the scala default method. In order to solve the scala problem, I migrated the scala sink interfaces(`AppendStreamTableSink`, `RetractStreamTableSink`, `UpsertStreamTableSink` etc) to java, thus the java subclasses do not have to inherit the default methods.


## Verifying this change

This change added tests and can be verified as follows:

  - Added integration tests, plan tests, validation tests for TableAggregate. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -1028,6 +1061,119 @@ class Table(
   }
 }
 
+class FlatAggTable(
```
I hesitate to remove `FlagAggTable`. 
- Even `FlatAggTable` and `GroupedFlatAggTable` share same methods now, I'm not sure if there is a chance that `FlatAggTable` will be different from `GroupedFlatAggTable` later. At that time, we should add `FlatAggTable` back which will break compatibility.
What do you think?

dianfu replied:
Can you describe when FlatAggTable may be needed? I'm fine to keep it if you're definitely sure it will be added back in the future. Otherwise I would suggest to remove it and only add it when necessary.

hequn8128 replied:
Ok, we can remove `FlatAggTable ` if we tend to only support select in `FlatAggTable ` and `GroupedFlatAggTable `.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -1028,6 +1061,119 @@ class Table(
+        Project(projectFields, table.logicalPlan).validate(table.tableEnv)
+      ).validate(table.tableEnv))
+
+    // check no '*' in the select of flatAggregate.
```
You are right. I want to make all flatAggregate(window, non-window) consistent in the first version. However, I can remove the check as you wish.

hequn8128 commented on flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/TableAggregateFunction.java:
```diff
@@ -0,0 +1,130 @@
+ * <pre>
+ * {@code
+ * Output data incrementally in upsert or append mode. For example, if we emit data for a TopN
+ * TableAggregateFunction, we don't have to output all top N elements each time a record comes.
```
have to -> need to ?

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/stream/table/TableAggregateTest.scala:
```diff
@@ -0,0 +1,159 @@
+import org.junit.Test
+
+class TableAggregateTest extends TableTestBase {
+
```
I have not supported alias since this will change `def flatAggregate(tableAggCall: TableAggFunctionCall)` to `def flatAggregate(tableAggCall: Expression)`. Maybe we can make `as` as a method for TableAggFunctionCall. This not only force a TableAggFunctionCall but also support alias?

hequn8128 commented:
@dianfu Thanks a lot for your review and great suggestions. 
I addressed almost all of your comments and updated the pr. Would be great if you can take another look.
Best, Hequn

dianfu commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/scala/StreamTableEnvironment.scala:
```diff
@@ -231,4 +231,20 @@ class StreamTableEnvironment(
+    * @tparam T The type of the output value.
+    * @tparam ACC The type of aggregate accumulator.
+    */
+  def registerFunction[T: TypeInformation, ACC: TypeInformation](
```
This method can be removed and reuse registerFunction(name: String, f: AggregateFunction[T, ACC])) by changing the signature to registerFunction(name: String, f: UserDefinedAggregateFunction[T, ACC]))

hequn8128 replied:
Yes, it would be good if we reuse it. However, It may not good to expose `UserDefinedAggregateFunction` to users. Besides, changing to `UserDefinedAggregateFunction` also makes it capable to register a class that extends directly from `UserDefinedAggregateFunction` which should be avoided.

dianfu replied:
Make sense.

dianfu commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/table/TableAggregateITCase.scala:
```diff
@@ -0,0 +1,227 @@
+  }
+
+  @Test(expected = classOf[ValidationException])
+  def testEmitMethod(): Unit = {
```
Move to TableAggregateValidationTest

hequn8128 replied:
It seems we can't move the test into `TableAggregateValidationTest` as the exception is thrown during translate to plan.

dianfu commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -800,6 +925,60 @@ object UserDefinedFunctionUtils {
+    // unwrap an Expression until we get a TableAggFunctionCall
+    def unwrap(expr: Expression): TableAggFunctionCall = expr match {
+      case Alias(child, name, extraNames: Seq[String]) =>
+        alias = Some(Seq(new UnresolvedFieldReference(name)) ++
```
new is unnecessary

dianfu commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/utils/UserDefinedFunctionUtils.scala:
```diff
@@ -800,6 +925,60 @@ object UserDefinedFunctionUtils {
+    def unwrap(expr: Expression): TableAggFunctionCall = expr match {
+      case Alias(child, name, extraNames: Seq[String]) =>
+        alias = Some(Seq(new UnresolvedFieldReference(name)) ++
+          extraNames.map(new UnresolvedFieldReference(_)))
```
new is unnecessary

dianfu commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamTableAggregateRule.scala:
```diff
@@ -0,0 +1,63 @@
+  val INSTANCE: DataStreamTableAggregateRule = new DataStreamTableAggregateRule()
+}
+
+
```
too many empty lines

hequn8128 commented:
@dianfu Thanks a lot for the review. I have addressed your comments and rebased to the latest master.

sunjincheng121 commented on flink-libraries/flink-table-common/src/main/java/org/apache/flink/table/functions/TableAggregateFunction.java:
```diff
@@ -0,0 +1,129 @@
+ * param: accumulator           the accumulator which contains the current aggregated results
+ * param: out                   the collector used to output data.
+ *
+ * public void emitValue(ACC accumulator, RetractableCollector out)
```
I think `emitValue` just like `getValue` always get complete full results. And `emitValueWithRetract` should get the incrementally result. Please see the detail example in [FLIP-29](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=97552739)

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala:
```diff
@@ -367,6 +367,38 @@ class Table(
+    *   tab.flatAggregate(tableAggFunc('a, 'b)).select('_1, '_2, '_3)
+    * }}}
+    */
+  def flatAggregate(tableAggCall: TableAggFunctionCall): FlatAggTable = {
```
`TableAggFunctionCall` is also an `Expression`. We don't need the check logic if we define the type as `TableAggFunctionCall`.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/aggregations.scala:
```diff
@@ -414,3 +414,98 @@ case class AggFunctionCall(
+  }
+}
+
+class TableAggFunctionCallAliasable(
```
`TableAggFunctionCallAliasable.as()` returns `TableAggFunctionCall`, so that users can't write multi as, such as `tab.flatAggregate(func('a, 'b) as ('a, 'b, c) as ('d, 'e, 'f))`

sunjincheng121 replied:
I meant that move the `as` method into `TableAggFunctionCall`.

hequn8128 replied:
What's the return type of `as` method? If it returns a `TableAggFunctionCall`. The `TableAggFunctionCall` can still perform another `as`. Users can write query like `tab.flatAggregate(func('a, 'b) as ('a, 'b, c) as ('d, 'e, 'f))`. 
I think we should avoid this invalid case.

sunjincheng121 replied:
Hmm,  also make sense to me due to `TableAggFunctionCallAliasable` is inner type. but I'm not sure, because table can be `as` many times. e.g.: `tab.as(..).as(..)`, I think  @twalthr and @fhueske can give us more useful suggestions.

hequn8128 commented on docs/dev/table/tableApi.md:
```diff
@@ -550,6 +550,82 @@ val result = orders.distinct()
+      <td>
+        <p>Similar to a <b>GroupBy Aggregation</b>. Groups the rows on the grouping keys with a following running table aggregation operator to aggregate rows group-wise. The difference from an AggregateFunction is that TableAggregateFunction may return 0 or more records for a group. You have to close the "flatAggregate" with a select statement. And the select statement does not support aggregate ...
+{% highlight scala %}
+val tableAggFunc: TableAggregateFunction = new MyTableAggregateFunction
```
I removed this line in `Table.scala` according to your suggestion. However, I think it would be better to keep this line in the document because it clearly indicates that tableAggFunc is a function of `TableAggregateFunction` which I think we be helpful for users.
What do you think?

hequn8128 replied:
Oh, my fault. I think it should be `val tableAggFunc: TableAggregateFunction[_, _] = new MyTableAggregateFunction`

hequn8128 commented on docs/dev/table/tableApi.md:
```diff
@@ -550,6 +550,82 @@ val result = orders.distinct()
+        <span class="label label-info">Result Updating</span>
+      </td>
+      <td>
+        <p>Similar to a <b>GroupBy Aggregation</b>. Groups the rows on the grouping keys with a following running table aggregation operator to aggregate rows group-wise. The difference from an AggregateFunction is that TableAggregateFunction may return 0 or more records for a group. You have to close the "flatAggregate" with a select statement. And the select statement does not support aggregate ...
```
How about: `select statement cannot contain aggregate expressions` ?
